File: docs/examples/Report.java
/* Report.java - the program that uses Stats.java.
 *
 * Keep both files in the same directory, then:
 *
 *   javac Report.java          # javac finds and compiles Stats.java too
 *   java Report
 *   java Report 4 8 15 16 23 42
 *
 * From https://aikaryashala.com/system_setup/05_install_java/
 */

public class Report {

    public static void main(String[] args) {
        int[] readings = parse(args);

        System.out.println("--- readings ---");
        System.out.printf("  count: %d%n", readings.length);
        System.out.printf("   mean: %.2f%n", Stats.mean(readings));
        System.out.printf("    max: %d%n", Stats.max(readings));
        System.out.printf("    min: %d%n", Stats.min(readings));
        System.out.printf(" spread: %d%n", Stats.spread(readings));
    }

    /** Use the numbers given on the command line, or a default set. */
    private static int[] parse(String[] args) {
        if (args.length == 0) {
            return new int[] {12, 7, 3, 21, 9, 15};
        }

        int[] numbers = new int[args.length];
        for (int i = 0; i < args.length; i++) {
            numbers[i] = Integer.parseInt(args[i]);
        }
        return numbers;
    }
}

File: docs/examples/report.py
"""report.py - the program to run under the debugger.

Keep this file and stats.py in the same directory, then:

    python3 report.py                 # run it - it crashes at the end, on purpose
    python3 -m pdb report.py          # walk through it a line at a time
    python3 -m pdb -c continue report.py   # let it crash, then inspect the wreckage

The crash at the bottom is deliberate. Finding out *why* it happens, without
adding a single print statement, is the point of the exercise.
"""

import stats

READINGS = [12, 7, 3, 21, 9, 15]


def show(title, numbers):
    summary = stats.summarise(numbers)
    print(f"--- {title} ---")
    for key, value in summary.items():
        print(f"{key:>7}: {value}")


def main():
    show("readings", READINGS)

    # This one is empty, and mean() divides by len(numbers).
    missing = []
    show("missing", missing)


if __name__ == "__main__":
    main()
